### ISSUE TITLE
Add context manager for invoke_later and invoke_when are useful for gui testing

### ISSUE THREAD

**Issue by kitchoi** [enthought/pyface#311]
I have learned, from a separate project that, three main functions are very useful for GUI testing.

(1) A context manager that schedules a function to be invoked later. The invocation will be canceled when leaving the context

(2) A context manager that schedules a function to be invoked when a condition is fulfilled.  The invocation will be canceled when leaving the context

(3) A function that starts the event loop when a condition is fulfilled, with a timeout.  This is already provided by `GuiTestAssistant.event_loop_until_condition`

(1) and (2) are not provided by pyface, and I had to write my own.

Then in tests, I do this pattern a lot:
```
with call_later(wait=0.1, function=run_something_that_makes_condition_true), \
            call_when(condition=condition, function=close_and_exit):
     loop_until(lambda: False, timeout=20.0)
```

The main idea about implementing these context managers is that they prevent test interactions:  An invocation scheduled in the event loop by one test will not never be run in another test.

**Comment by mdickinson** [comment]
Sounds good to me. Our aim should be to provide a consistent set of tools that can be used in all projects that need GUI testing (and there should be One Obvious Way To Do It). That may mean combining the proposed context managers with the GuiTestAssistant somehow.

**Comment by corranwebster** [comment]
I agree that this sounds like a good idea.  Possibly the right place for the context managers is with the `GUI` object which also has the `invoke_later` and similar methods?
